src/bootstrap/bootstrap.py

```diff
@@ -533,7 +533,7 @@ def download_toolchain(self):
         bin_root = self.bin_root()
 
         key = self.stage0_compiler.date
-        if self.rustc().startswith(bin_root) and \
+        if (self.rustc().startswith(bin_root) or self.cargo().startswith(bin_root)) and \
```

Comment:
This logic breaks if you do this:
1) Set `build.rustc`
2) `x build` (downloads just rustc)
3) Unset `build.rustc`, set `build.cargo`
4) `x build` -> fails, because cargo should be downloaded but it isn't, because the stage0 rustc is present on disk

It's a bit of an edge, but wanted to let you know nevertheless.

Reply:
It was the opposite for me (setting `build.cargo` first, and then `build.rustc`), which should be covered with [this change](https://github.com/rust-lang/rust/compare/98fc12d40a7f64b1ec84c16505e4fded9f5c713a..569e0e10ad825db5d8f5c530cd9eda026404f2aa). Thanks!

Reply:
Yeah, I got the order wrong 🤦 I guess that there is now still an edge case; when you have an old cargo stage0 on disk, and there has been a beta bump in the meantime, the new cargo might not get downloaded (before cargo would get redownloaded automatically because the rustc stamp would change). But this would manifest only if you use build.rustc without build.cargo and you don't clear the build directory after a beta bump.

Reply:
> so I think that we can just ignore it

We just need to call `program_out_of_date` for cargo, [just done that](https://github.com/rust-lang/rust/compare/569e0e10ad825db5d8f5c530cd9eda026404f2aa..5f2cedc5dc4ed5653069a83f1ed98b46f0c33c7c) as it's easy to handle/cover.

Reply:
Ah, wasn't sure if it's indeed that easy. T hanks!